oauth2client/contrib/gce.py

```diff
@@ -85,32 +121,110 @@ class AppAssertionCredentials(AssertionCredentials):
     information to generate and refresh its own access tokens.
     """
 
-    @util.positional(2)
-    def __init__(self, scope='', **kwargs):
+    @util.positional(3)
+    def __init__(self,
+                 scope=None,
+                 service_account_email='default',
+                 service_account_info=None,
+                 **kwargs):
         """Constructor for AppAssertionCredentials
 
         Args:
-            scope: string or iterable of strings, scope(s) of the credentials
-                   being requested. Using this argument will have no effect on
-                   the actual scopes for tokens requested. These scopes are
-                   set at VM instance creation time and won't change.
+            scope:
+                string or iterable of strings, scope(s) of the credentials
+                being requested. Using this argument will have no effect on
+                the actual scopes for tokens requested. These scopes are
+                set at VM instance creation time and won't change.
+            service_account_email:
+                the email for these credentials. This can be used with custom
+                service accounts, or left blank to use the default service
+                account for the instance. Usually the compute engine service
+                account.
+            service_account_info:
+                Deserialized JSON object, returned by self.service_account_info
         """
+
+        self._service_account_info = service_account_info or {
+            'email': service_account_email
+        }
+        self._project_id = None
+        self._partial = service_account_info is None
+
         if scope:
-            warnings.warn(_SCOPES_WARNING)
-        # This is just provided for backwards compatibility, but is not
-        # used by this class.
-        self.scope = util.scopes_to_string(scope)
+            if self.has_scopes(scope):
+                warnings.warn(_SCOPES_WARNING)
+            else:
+                raise ValueError(_SCOPES_WARNING)
```

Comment:
Yeah, I'd like to fail faster when a user instantiates these credentials with scopes it cannot possibly have. e.g. a user calls `credentials.create_scoped(['a','b'])` and these credentials only have scope `'c'`. Calls which require scopes `'a'` and `'b'` will fail *when they are made* but we know that these credentials don't match user expectations when they are instantiated so we should fail then. 

In the case that the user tries to create these credentials with scopes the *do* have we can just warn them that they shouldn't be specifying scopes, but the credentials will work as expected.